### PR TITLE
Incr. comp. dep-node for traits, tests

### LINKED ISSUE #33850: struct field dependencies are overly conservative
The test cases I am adding in an upcoming PR show that for some reason we consider functions as dirty when any struct changes, even if the fn doesn't use that struct.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Introduce new tests and also make dep-node for trait selection a bit more selective.

Fixes #33850

r? @michaelwoerister

### CODE REVIEW

File: src/librustc/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -212,7 +221,11 @@ impl<D: Clone + Debug> DepNode<D> {
             TraitImpls(ref d) => op(d).map(TraitImpls),
             TraitItems(ref d) => op(d).map(TraitItems),
             ReprHints(ref d) => op(d).map(ReprHints),
-            TraitSelect(ref d) => op(d).map(TraitSelect),
+            TraitSelect(ref d, ref type_ds) => {
+                let d = try_opt!(op(d));
+                let type_ds = try_opt!(type_ds.iter().map(|d| op(d)).collect());
```
Comment (michaelwoerister) [inline src/librustc/dep_graph/dep_node.rs]: I don't understand why this works. Doesn't the argument to `try_opt!()` have to be an `Option` of something? But `collect()` will return a `Vec`?

File: src/librustc/dep_graph/raii.rs [hunk 2]
```diff
@@ -20,14 +20,14 @@ pub struct DepTask<'graph> {
...
-        data.enqueue(DepMessage::PushTask(key));
+        data.enqueue(DepMessage::PushTask(key.clone()));
         DepTask { data: data, key: key }
     }
 }
 
 impl<'graph> Drop for DepTask<'graph> {
     fn drop(&mut self) {
-        self.data.enqueue(DepMessage::PopTask(self.key));
+        self.data.enqueue(DepMessage::PopTask(self.key.clone()));
```
Comment (michaelwoerister) [inline src/librustc/dep_graph/raii.rs]: This `clone()` could be avoided by swapping with an empty DepNode maybe.

File: src/librustc/dep_graph/dep_node.rs (hunk 1 above)
Comment (jonas-schievink) [inline src/librustc/dep_graph/dep_node.rs]: You can collect an iterator yielding `Option`s into a single `Option`

File: src/librustc/ty/mod.rs [hunk 3]
```diff
@@ -946,7 +946,28 @@ impl<'tcx> TraitPredicate<'tcx> {
...
     fn dep_node(&self) -> DepNode<DefId> {
-        DepNode::TraitSelect(self.def_id(), vec![])
+        // Ideally, the dep-node would just have all the input types
+        // in it.  But they are limited to including def-ids. So as an
+        // approximation we include the def-ids for all nominal types
+        // found somewhere. This means that we will e.g. conflate the
+        // dep-nodes for `u32: SomeTrait` and `u64: SomeTrait`, but we
+        // would have distinct dep-nodes for `Vec<u32>: SomeTrait`,
+        // `Rc<u32>: SomeTrait`, and `(Vec<u32>, Rc<u32>): SomeTrait`.
+        // Note that it's always sound to conflate dep-nodes, it jus
```
Comment (michaelwoerister) [inline src/librustc/ty/mod.rs]: jus -> just

File: src/librustc/dep_graph/dep_node.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc/dep_graph/dep_node.rs]: @jonas-schievink Oh, that's interesting. Thanks for enlightening me `:)`

**Comment by michaelwoerister** [comment]
LGTM! r=me with the typo fixed. I'll leave it up to you, if you want to get rid of the avoidable allocation in `DepTask::drop`. It's not occurring in a tight loop exactly.

File: src/librustc/dep_graph/dep_node.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc/dep_graph/dep_node.rs]: > You can collect an iterator yielding Options into a single Option

best feature ever

**Pull request merged**